api/src/convert/tx_in.rs

```diff
@@ -89,9 +102,99 @@ impl TryFrom<&external::InputRules> for InputRules {
             .map(tx::TxOut::try_from)
             .collect::<Result<Vec<_>, _>>()?;
         let max_tombstone_block = source.max_tombstone_block;
+        let fractional_outputs = source
+            .fractional_outputs
+            .iter()
+            .map(RevealedTxOut::try_from)
+            .collect::<Result<Vec<_>, _>>()?;
+        let fractional_change = source
+            .fractional_change
+            .as_ref()
+            .map(RevealedTxOut::try_from)
+            .transpose()?;
+        let max_allowed_change_value = source.max_allowed_change_value;
         Ok(InputRules {
             required_outputs,
             max_tombstone_block,
+            fractional_outputs,
+            fractional_change,
+            max_allowed_change_value,
         })
     }
 }
+
+/// Convert RevealedTxOut --> external::RevealedTxOut.
+impl From<&RevealedTxOut> for external::RevealedTxOut {
+    fn from(source: &RevealedTxOut) -> Self {
+        let mut result = external::RevealedTxOut::new();
+        result.set_tx_out(external::TxOut::from(&source.tx_out));
+        result.set_amount_shared_secret(source.amount_shared_secret.to_vec());
+        result
+    }
+}
+
+/// Convert external::RevealedTxOut --> RevealedTxOut
+impl TryFrom<&external::RevealedTxOut> for RevealedTxOut {
+    type Error = ConversionError;
+
+    fn try_from(source: &external::RevealedTxOut) -> Result<Self, Self::Error> {
+        let tx_out =
+            tx::TxOut::try_from(source.tx_out.as_ref().ok_or(Self::Error::ObjectMissing)?)?;
+        let amount_shared_secret = source.get_amount_shared_secret().to_vec();
+        Ok(RevealedTxOut {
+            tx_out,
+            amount_shared_secret,
+        })
```

Comment:
I am not sure.

We have code elsewhere in the functions that use `RevealedTxOut` that would surface this error. So we could treat that as a business logic bug rather than a serialization bug.

In a lot of existing code I think we do flag these immediately.

But I think we have been trying to embrace a new world where we have distinct Proto types and business logic types?

I think it can't hurt to flag this immediately, and since a bunch of our existing code does that I should probably continue the pattern.